src/librustc/ty/query/job.rs

```diff
@@ -30,61 +31,119 @@ pub struct QueryInfo<'tcx> {
     pub query: Query<'tcx>,
 }
 
-/// Representss an object representing an active query job.
-pub struct QueryJob<'tcx> {
+type QueryMap<'tcx> = FxHashMap<QueryToken, QueryJobInfo<'tcx>>;
+
+/// A value uniquely identifiying an active query job.
+/// This value is created from a stack pointer in `get_query` and `force_query`
+/// which is alive while the query executes.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct QueryToken(NonZeroUsize);
```

Comment:
Using a pointer into the stack as a unique identifier is clever but I'm a bit worried that we can't make the compiler guarantee that the value won't be reused. The core logic within the query engine is very complicated and somebody might do some kind of optimization that silently breaks the invariant of the `key` stack value being alive throughout the query jobs lifetime.

If I understand things correctly, the requirements for `QueryToken` are:
 - it needs to uniquely identify a QueryJob
 - it needs to be cheap to construct.

At construction time we already have a lock on the query cache. Could we put a `u64` counter into that?

Also, if the main purpose of `QueryToken` is to identify a `QueryJob` then I'd prefer if it was called `QueryJobId`.

Reply:
We can put a counter into the query cache shard. We'll need to split `QueryToken` into `QueryJobId` (which globally identifies a query) and `QueryShardJobId` (which just identifies a query within the shard), or we could just use `u32` for the latter.

Reply:
I was thinking just having a u64 counter per shard and initializing each counter in a way that the upper N bits globally identify the shard. E.g. allocate a 20 bit "(query-kind, shard) id" from a static atomic counter when the shard is created and then initialize the shard-local counters with `shard_id << 44`. But I don't actually care about the implementation details as long as `QueryJobId` does what its name suggests.

Reply:
I used `DepKind`, `u16` (for the shard) and `u32` for a job in the shard. We're limited to `u32::MAX - 1` due to `DepNodeIndex` anyway.

Reply:
Yes, I like how that turned out.